src/state_modeler.rs

```diff
@@ -0,0 +1,269 @@
+/// pressure data by calling this struct with the `query` function
+#[doc = include_str!("../docs/notations.html")]
+#[doc = include_str!("../docs/stylus_state_modeler.html")]
+pub struct StateModeler {
+    /// max number of elements
+    stylus_state_modeler_max_input_samples: usize,
+    /// deque holding the data from strokes
+    last_strokes: VecDeque<ModelerInput>,
+}
+
+impl Default for StateModeler {
+    fn default() -> Self {
+        Self {
+            stylus_state_modeler_max_input_samples: 10,
+            last_strokes: VecDeque::with_capacity(11),
+        }
+    }
+}
+
+impl StateModeler {
+    /// initialize a new StateModeler
+    pub fn new(param: usize) -> Self {
+        Self {
+            stylus_state_modeler_max_input_samples: param,
+            last_strokes: VecDeque::with_capacity(param + 1),
+        }
+    }
+
+    /// add the most recent raw input to the StateModeler
+    pub fn update(&mut self, input: ModelerInput) {
+        // add the event to the strokes
+        self.last_strokes.push_back(input);
+        if self.last_strokes.len() > self.stylus_state_modeler_max_input_samples {
+            self.last_strokes.pop_front();
+        }
+    }
+
+    /// reset the StateModeler
+    pub fn reset(&mut self, max_input: usize) {
+        self.last_strokes = VecDeque::new();
+        self.stylus_state_modeler_max_input_samples = max_input;
+    }
+
+    /// query the pressure by interpolating it from raw input events
+    pub fn query(&mut self, pos: (f32, f32)) -> f32 {
+        // iterate over the decque
+        match self.last_strokes.len() {
+            0 => 1.0,
+            1 => return self.last_strokes.front().unwrap().pressure,
+            _ => {
+                let mut distance = f32::INFINITY;
+                let mut r: f32 = 0.0;
+
+                let mut start_pressure: f32 = 1.0;
+                let mut end_pressure: f32 = 1.0;
+
+                for index_it in 0..self.last_strokes.len() - 1 {
+                    let start_pos = self.last_strokes.get(index_it).unwrap().pos;
+                    let end_pos = self.last_strokes.get(index_it + 1).unwrap().pos;
+
+                    let r_c = nearest_point_on_segment(start_pos, end_pos, pos);
+                    let point_c = interp2(start_pos, end_pos, r_c);
+
+                    if dist(pos, point_c) < distance {
+                        distance = dist(pos, point_c);
+                        r = r_c;
+                        start_pressure = self.last_strokes.get(index_it).unwrap().pressure;
+                        end_pressure = self.last_strokes.get(index_it + 1).unwrap().pressure;
+                    }
+                }
+
+                interp(start_pressure, end_pressure, r)
+            }
+        }
+    }
+}
+
+#[test]
+fn state_modeler_straight() {
+    let mut state_mod = StateModeler::new(10);
+    assert_eq!(state_mod.query((0.0, 0.0)), 1.0); // 1 is our "unknown" default value
```

Comment:
It would be better to use `approx-eq`'s `assert_rel_eq` here as well and in all other occurrences of float compares